internal/api/courses/admin/email_test.go

```diff
@@ -0,0 +1,166 @@
+package admin
+
+import (
+	"testing"
+
+	"github.com/edulinq/autograder/internal/api/core"
+	"github.com/edulinq/autograder/internal/email"
+	"github.com/edulinq/autograder/internal/util"
+)
+
+func TestEmail(test *testing.T) {
+	defer email.ClearTestMessages()
+
+	testCases := []struct {
+		UserEmail string
+		Message   email.Message
+		Locator   string
+		DryRun    bool
+		PermError bool
+	}{
+		// No subject.
+		{
+			UserEmail: "course-grader",
+			Message: email.Message{
+				To: []string{"*"},
+			},
+			Locator: "627",
+		},
+
+		// No recipients.
+		{
+			UserEmail: "course-grader",
+			Message: email.Message{
+				Subject: "Message in a Bottle",
+			},
+			Locator: "-628",
+		},
+
+		// Invalid permissions.
+		{
+			UserEmail: "course-other",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email from Course Student",
+			},
+			Locator:   "-020",
+			PermError: true,
+		},
+		{
+			UserEmail: "course-student",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email from Course Student",
+			},
+			Locator:   "-020",
+			PermError: true,
+		},
+
+		// Invalid permissions, role escalation.
+		{
+			UserEmail: "server-user",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email from Course Student",
+			},
+			Locator:   "-040",
+			PermError: true,
+		},
+		{
+			UserEmail: "server-creator",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email from Course Student",
+			},
+			Locator:   "-040",
+			PermError: true,
+		},
+
+		// Valid permissions.
+		{
+			UserEmail: "course-grader",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+		},
+		{
+			UserEmail: "course-admin",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+		},
+		{
+			UserEmail: "course-owner",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+		},
+
+		// Valid permissions, role escalation.
+		{
+			UserEmail: "server-admin",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+		},
+		{
+			UserEmail: "server-owner",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+		},
+
+		// Dry run.
+		{
+			UserEmail: "course-grader",
+			Message: email.Message{
+				To:      []string{"*"},
+				Subject: "Email for All Course Users",
+			},
+			DryRun: true,
+		},
```

Comment:
Missing some test cases.
E.g.,
 - To: course-student (role), -course-student@test.edulinq.org
 - CC
 - BCC
 - Body
 - HTML